include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -1491,9 +1494,20 @@ namespace internal
         }
     }
 
-    const bool hp_functionality_enabled =
-      std::any_of(dof_handlers.begin(), dof_handlers.end(), [](const auto &dh) {
-        return dh->has_hp_capabilities();
+    const auto [hp_functionality_enabled, first_dof_handler_index] =
+      std::invoke([&dof_handlers]() -> std::pair<bool, unsigned int> {
```

Comment:
Can you explain why you need to compute `first_dof_handler_index` again at this point? This should be called by `internal_reinit()` (via `initialize_dof_indices`), that gets called after you set the class variable?

Reply:
This is not a member function of MatrixFree, so either the value must be passed to this function or computed within. Here, the bool `hp_functionality_enabled` is determined, and we can determine `first_dof_handler_index` as a by-product. I chose this way to avoid adding another argument to this function. Do you prefer to pass `first_hp_dof_handler_index` to this function instead?

Reply:
Ah, I didn't see that this is not a member function. It's perfectly fine in that case, let's keep your version. :+1: